kotlin/lib/src/main/kotlin/truelayer/signing/Jws.kt

```diff
@@ -0,0 +1,107 @@
+package truelayer.signing
+
+import com.nimbusds.jose.*
+import com.nimbusds.jose.crypto.ECDSASigner
+import com.nimbusds.jose.crypto.ECDSAVerifier
+import com.nimbusds.jose.util.Base64URL
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import java.security.interfaces.ECPrivateKey
+import java.security.interfaces.ECPublicKey
+
+internal fun signEs512(
+    key: ECPrivateKey,
+    kid: String,
+    headers: Map<HeaderName, String>,
+    method: String,
+    path: String,
+    body: ByteArray
+): Result<String> =
+    InvalidSignatureException.evaluate {
+        val json = Json.encodeToString(
+            mapOf(
+                "alg" to "ES512",
+                "kid" to kid,
+                "tl_version" to "2",
+                "tl_headers" to headers.keys.joinToString { it.name }
+            )
+        ).toByteArray()
+
+        val jwsObject = JWSObject(
+            JWSHeader.parse(Base64URL(json.toUrlBase64())),
+            Payload(Base64URL(buildPayload(headers, method, path, body)))
+        )
+
+        jwsObject.sign(ECDSASigner(key))
+        jwsObject.serialize(true)
+    }
+
+internal fun verifyTlSignature(
+    signature: String,
+    publicKey: ECPublicKey,
+    requiredHeaders: HashSet<String>,
+    method: String,
+    path: String,
+    body: ByteArray,
+    headers: Map<HeaderName, String>
+): Result<Boolean> {
```

Comment:
Boolean is not a great design choice here, but it is forced by the library we use to verify